map/mode.c

```diff
@@ -7,69 +7,367 @@
+    ta_log_error("%d\n", ret);
+  }
   return ret;
 }
 
-retcode_t map_announce_endpoint(mam_api_t *const api, tryte_t const *const channel_id,
-                                bundle_transactions_t *const bundle, trit_t *const msg_id,
-                                tryte_t *const new_endpoint_id) {
+/**
+ * Writes a packet on a bundle
+ *
+ * @param api - The API [in,out]
+ * @param bundle - The bundle that the packet will be written into [out]
+ * @param payload - The payload to write [in]
+ * @param msg_id - The msg_id [in]
+ *
+ * @return return code
+ */
+static retcode_t map_write_packet(mam_api_t *const api, bundle_transactions_t *const bundle, char const *const payload,
+                                  trit_t const *const msg_id) {
   retcode_t ret = RC_OK;
+  const bool last_packet = true;
+  const mam_msg_checksum_t checksum = MAM_MSG_CHECKSUM_SIG;
+  tryte_t *payload_trytes = (tryte_t *)malloc(2 * strlen(payload) * sizeof(tryte_t));
 
-  ERR_BIND_RETURN(mam_api_endpoint_create(api, MSS_DEPTH, channel_id, new_endpoint_id), ret);
+  ascii_to_trytes(payload, payload_trytes);
 
-  ret = mam_api_bundle_announce_endpoint(api, channel_id, new_endpoint_id, NULL, NULL, bundle, msg_id);
+  ret = mam_api_bundle_write_packet(api, msg_id, payload_trytes, strlen(payload) * 2, checksum, last_packet, bundle);
 
+  free(payload_trytes);
   return ret;
 }
 
-retcode_t map_write_header_on_channel(mam_api_t *const api, tryte_t const *const channel_id,
-                                      bundle_transactions_t *const bundle, trit_t *const msg_id) {
-  retcode_t ret = RC_OK;
+/**
+ * Find wheter the tag of the first transaction of the bundle already has existed on Tangle, which means this current
+ * Header has been published.
+ *
+ * @param tag_array - An tag array contains all the tag under the given Channel ID[in,out]
+ * @param bundle - The bundle contains the Header may be used[out]
+ *
+ * @return Return true is the tags exist in tag array
+ */
+static bool msg_id_exist(hash81_array_p tag_array, bundle_transactions_t **bundle) {
+  bool existed = false;
+  flex_trit_t *tag = transaction_tag((iota_transaction_t *)utarray_front(*bundle));
+  tryte_t tag_tryte[NUM_TRYTES_TAG + 1], tag_cursor_tryte[NUM_TRYTES_TAG + 1];
+  tag_tryte[NUM_TRYTES_TAG] = 0;
+  tag_cursor_tryte[NUM_TRYTES_TAG] = 0;
 
-  ret = mam_api_bundle_write_header_on_channel(api, channel_id, NULL, NULL, bundle, msg_id);
+  flex_trits_to_trytes(tag_tryte, NUM_TRYTES_TAG, tag, NUM_FLEX_TRITS_TAG, NUM_FLEX_TRITS_TAG);
 
-  return ret;
+  flex_trit_t *tag_cursor = NULL;
+  for (tag_cursor = (flex_trit_t *)utarray_front(tag_array); tag_cursor != NULL;
+       tag_cursor = (flex_trit_t *)utarray_next(tag_array, tag_cursor)) {
+    flex_trits_to_trytes(tag_cursor_tryte, NUM_TRYTES_TAG, tag_cursor, NUM_FLEX_TRITS_TAG, NUM_FLEX_TRITS_TAG);
+    if (!memcmp(tag_cursor_tryte, tag_tryte, NUM_TRYTES_TAG)) {
+      existed = true;
+      break;
+    }
+  }
+
+  bundle_transactions_free(bundle);
```

Comment:
I have no idea why you perform `new` right after `free`.

Reply:
With doing so, the `bundle` is initialized, so it can be reuse.

Reply:
I dislike the way. Instead, can you consider the concept of [reference counting](https://en.wikipedia.org/wiki/Reference_counting)?

Reply:
How about I move the re-initialization function calls from `msg_id_exist` to the for loop (about line 300).
The bundle isn't referenced multiple times and in a complex way as well, so use reference counting may be overkill

Reply:
> How about I move the re-initialization function calls from `msg_id_exist` to the for loop (about line 300).

It sounds reasonable.